src/Symfony/Component/Security/Core/Role/RoleHierarchy.php

```diff
@@ -65,9 +65,10 @@ protected function buildRoleMap()
                 }
 
                 $visited[] = $role;
-                $this->map[$main] = array_unique(array_merge($this->map[$main], $this->hierarchy[$role]));
+                $this->map[$main] = array_merge($this->map[$main], $this->hierarchy[$role]);
                 $additionalRoles = array_merge($additionalRoles, array_diff($this->hierarchy[$role], $visited));
             }
+            $this->map[$main] = array_unique($this->map[$main]);
         }
```

Comment:
`call_user_func_array('array_merge', $this->map)` would be the 5.5 equivalent, provided `$maps = array($this->map[$main]);` before the loop